designs/webviews/webviews-general.md

```diff
@@ -0,0 +1,95 @@
+
+## Proposed Solution
+
+I propose we move more complex UIs to VS Code webviews, with "more complex" defined as:
+
+* UIs with multiple inputs that could invalidate or change other data
+  * For instance: changing an AWS region, which changes the values present in select boxes
+* UIs where users don't have a reason to act in a linear fashion
+  * For instance: a form wrapping an AWS service call, where a user would want to enter ARNs from multiple locations
+* UIs that have more complex inputs, e.g. arbitrary file paths (especially to support drag-and-drop), JSON, etc.
+* Forms that could feasibly be submitted more than once, e.g. Lambda invocations
+* UIs that involve paginated loading and filtering from the Node backend
+* Forms that could benefit from being displayed next to something else, instead of dominating focus
+* UIs with arbitrary, non-form-like functionality
+  * For instance: UIs that display read-only data
+* UIs involving more than 3 steps (or some other arbitrary number to denote complexity)
+
+We should not abandon the use of VS Code UI constructs. VS Code constructs excel at quickly inputting small amounts of data, especially if it's in a fairly linear fashion, without much need for backtracking. Additionally, VS Code offers a fully-featured tree view interface; unless this absolutely requires integration into a workflow, we should keep tree views within the IDE.
+
+As an example, based on these rules, I would propose we do the following with our current workflows:
+
+| Workflow Name                        | Move to Webview? | Why? |
+| ------------------------------------ | ---------------- | ---- |
+| Connect to AWS                       | No               | Selecting a single option from a list. |
+| Show region in the explorer          | No               | Selecting a single option from a list. |
+| Hide region in the explorer          | No               | Selecting a single option from a list. |
+| Create credential profile            | No               | Opens an existing file for editing with a link to documentation and involves no connection to a backend. |
+| Create SAM Application               | Yes              | Brings us closer to a form-like interface and allows us to add more config options (e.g. the Jetbrains Create workflow). |
+| Configure SAM Application (codelens) | Yes              | Gives us a true interface to configure local invocations instead of a confusing JSON file. Experience could be made similar to the Cloud9 and Jetbrains local invokers. |
+| Deploy SAM Application               | Yes              | Would be a more logical flow when populating data based on region. Users can change region and invalidate other fields, and refresh potential selections. |
+
+VS Code webviews are not a cureall, and introduce a new set of complexities. A list, along with mitigations, follows:
+
+* Webviews are HTML + JS running in an `iframe`, which adds a considerable resource burden on the IDE (see Appendix 1).
```

Comment:
when you say 'burden' my mind immediately jumps to being more resource intensive (e.g. memory hog, 100% cpu etc) - but the reasons that are listed in the VS Code documentation don't really elude to this - their caution is more based on the usability.